src/client/ds3/bucket.c

```diff
@@ -0,0 +1,419 @@
+		rc = daos_der2errno(rc);
+		goto err;
+	}
+
+	for (int i = 0; i < ncont; i++) {
+		char *name = conts[i].pci_label;
+		if (strcmp(name, METADATA_BUCKET) == 0) {
+			D_INFO("Skipping container: %s", name);
+			continue;
+		}
+
+		// Copy bucket name
+		strcpy(buf[bi].name, name);
+
+		// Get info
+		rc = ds3_bucket_open(name, &ds3b, ds3, ev);
+		if (rc != 0) {
+			D_INFO("Skipping container: %s", name);
+			continue;
+		}
+
+		rc = ds3_bucket_get_info(&buf[bi], ds3b, ev);
+		if (rc != 0) {
+			D_INFO("Skipping container: %s", name);
+			rc = ds3_bucket_close(ds3b, ev);
+			continue;
+		}
+
+		rc = ds3_bucket_close(ds3b, ev);
+		if (rc != 0)
+			goto err;
+
+		bi++;
+		// TODO handle marker
+	}
+
+	*nbuck = bi;
+
+err:
+	D_FREE(conts);
+	return -rc;
+}
+
+int
+ds3_bucket_create(const char *name, struct ds3_bucket_info *info, dfs_attr_t *attr, ds3_t *ds3,
+		  daos_event_t *ev)
+{
+	int           rc   = 0;
+	ds3_bucket_t *ds3b = NULL;
+
+	if (ds3 == NULL || name == NULL)
+		return -EINVAL;
+
+	// Prevent attempting to create metadata bucket
+	if (strcmp(name, METADATA_BUCKET) == 0) {
+		D_ERROR("Cannot create metadata bucket");
+		return -EINVAL;
+	}
+
+	// Create dfs container and open ds3b
+	rc = dfs_cont_create_with_label(ds3->poh, name, attr, NULL, NULL, NULL);
+	if (rc != 0) {
+		D_ERROR("Failed to create container, rc = %d\n", rc);
+		return -rc;
+	}
+
+	rc = ds3_bucket_open(name, &ds3b, ds3, ev);
+	if (rc != 0) {
+		D_ERROR("Failed to open container, rc = %d\n", rc);
+		return -rc;
+	}
+
+	rc = ds3_bucket_set_info(info, ds3b, ev);
+	if (rc != 0) {
+		D_ERROR("Failed to put bucket info, rc = %d\n", rc);
+		goto err;
+	}
+
+	// Create multipart index
+	rc = dfs_mkdir(ds3->meta_dfs, ds3->meta_dirs[MULTIPART_DIR], name, DEFFILEMODE, 0);
+	if (rc != 0 && rc != EEXIST) {
+		D_ERROR("Failed to create multipart index, rc = %d\n", rc);
+	}
+
+err:
+	ds3_bucket_close(ds3b, ev);
+	return -rc;
+}
+
+int
+ds3_bucket_destroy(const char *name, bool force, ds3_t *ds3, daos_event_t *ev)
+{
+	int            rc      = 0;
+	ds3_bucket_t  *ds3b    = NULL;
+	uint32_t       nd      = 10;
+	struct dirent *dirents = NULL;
+	dfs_obj_t     *dir_obj = NULL;
+	daos_anchor_t  anchor;
+
+	if (ds3 == NULL || name == NULL)
+		return -EINVAL;
+
+	rc = ds3_bucket_open(name, &ds3b, ds3, ev);
+	if (rc != 0) {
+		return -rc;
+	}
+
+	// Check if the bucket is empty
+	if (!force) {
+		rc = dfs_lookup(ds3b->dfs, "/", O_RDWR, &dir_obj, NULL, NULL);
+		if (rc != 0) {
+			goto err_ds3b;
+		}
+
+		D_ALLOC_ARRAY(dirents, nd);
+		if (dirents == NULL) {
+			rc = ENOMEM;
+			goto err_dir_obj;
+		}
+
+		daos_anchor_init(&anchor, 0);
+		rc = dfs_readdir(ds3b->dfs, dir_obj, &anchor, &nd, dirents);
+		if (rc != 0) {
+			goto err_dirents;
+		}
+
+		// The bucket is not empty
+		if (nd != 0) {
+			rc = ENOTEMPTY;
+			goto err_dirents;
+		}
+	}
+
+	// Remove the bucket's multipart directory
+	rc = dfs_remove(ds3->meta_dfs, ds3->meta_dirs[MULTIPART_DIR], name, true, NULL);
```

Comment:
is it ok to orphan all the directory contents? or do you reference them and clean them up somewhere else?

Reply:
yes indeed.